### PR TITLE
Cfy 1879

### PR DESCRIPTION
**Pull request by EarthmanT**
CFY-1879
add non-required server/instance attribute/parameter dict to ec2.instance.create
Also added logging.

### CODE REVIEW

File: ec2/instance.py [hunk 1]
```diff
@@ -130,16 +216,26 @@ def _state_validation(instance_id, state, timeout_len):
...
+        ctx.logger.debug('Instance is running.')
     else:
         raise NonRecoverableError('Instance not running.')
+
+
+def build_arg_dict(user_supplied, unsupported):
+
+    arguments = {}
+    for pair in user_supplied.items():
+        arguments['{0}'.format(pair[0])] = pair[1]
```
Comment (EarthmanT) [inline ec2/instance.py]: it worked!

File: ec2/instance.py [hunk 2]
```diff
@@ -38,28 +38,56 @@
...
+    """ Creates an AWS Instance from an (AMI) image_id and an instance_type.
     """
-    :return: reservation object
-    """
 
-    ami_image_id = ctx.node.properties['ami_image_id']
-    instance_type = ctx.node.properties['instance_type']
+    arguments = build_arg_dict(ctx.node.properties['attributes'].copy(),
+                               RUN_INSTANCES_UNSUPPORTED)
+    arguments['image_id'] = ctx.node.properties['image_id']
```
Comment (EarthmanT) [inline ec2/instance.py]: no prob bob!

File: ec2/tests/test_ec2.py [hunk 3]
```diff
@@ -26,18 +25,21 @@
...
-RESERVATION_OBJECT_TYPE = Reservation
 
 
 class TestPlugin(unittest.TestCase):
 
-    def test_instance_create(self):
+    def mock_ctx(self, test_name):
 
-        test_name = 'test_instance_create'
         test_node_id = '{0}'.format(test_name)
```
Comment (EarthmanT) [inline ec2/tests/test_ec2.py]: done

File: ec2/instance.py [hunk 4]
```diff
@@ -107,21 +173,41 @@ def terminate(**kwargs):
...
+
+    if _state_validation(instance_id, INSTANCE_TERMINATED, TERMINATION_TIMEOUT, CHECK_INTERVAL):
+        ctx.logger.info('(Node: {0}): Instance stopped. (Instance id: {1}).'
+                        .format(ctx.instance.id, instance_id))
     else:
-        raise RecoverableError('Instance did not terminate within specified timeout: {0}.'
-                               .format(TERMINATION_TIMEOUT))
+        ctx.logger.error('(Node: {0}): Failed to verify that the instance is terminated. (Instance id: {1}).'
+                         .format(ctx.instance.id, instance_id))
+        del ctx.instance.runtime_properties['instance_id']
```
Comment (EarthmanT) [inline ec2/instance.py]: moved

File: ec2/instance.py [hunk 5]
```diff
@@ -107,21 +173,41 @@ def terminate(**kwargs):
...
+def _state_validation(instance_id, state, timeout_length, check_interval):
 
-def _state_validation(instance_id, state, timeout_len):
+    ctx.logger.debug('Beginning state validation: instance id: {0}, state: {1}, timeout length: {2}')
 
-    timeout = time.time() + timeout_len
+    timeout = time.time() + timeout_length
 
     while True:
         instance_state = EC2().get_all_instance_status(instance_ids=instance_id)[0]
```
Comment (EarthmanT) [inline ec2/instance.py]: added an operation to validate the instance id refers to an actual instance and only then validate the state

File: ec2/instance.py [hunk 6]
```diff
@@ -107,21 +173,41 @@ def terminate(**kwargs):
...
-        raise RecoverableError('Instance did not terminate within specified timeout: {0}.'
-                               .format(TERMINATION_TIMEOUT))
+        ctx.logger.error('(Node: {0}): Failed to verify that the instance is terminated. (Instance id: {1}).'
+                         .format(ctx.instance.id, instance_id))
+        del ctx.instance.runtime_properties['instance_id']
+        raise NonRecoverableError('(Node: {0}): Instance did not terminate within specified timeout: {0}.'
+                                  .format(ctx.instance.id, TERMINATION_TIMEOUT))
+
 
+def _state_validation(instance_id, state, timeout_length, check_interval):
```
Comment (EarthmanT) [inline ec2/instance.py]: decided against this because we're really only interested in the end state

File: ec2/instance.py [hunk 7]
```diff
@@ -107,21 +173,41 @@ def terminate(**kwargs):
...
 
-def _state_validation(instance_id, state, timeout_len):
+    ctx.logger.debug('Beginning state validation: instance id: {0}, state: {1}, timeout length: {2}')
```
Comment (EarthmanT) [inline ec2/instance.py]: added this to logging

File: ec2/instance.py [hunk 8]
```diff
@@ -38,28 +38,56 @@
...
+        reservation = EC2().run_instances(**arguments)
     except EC2ResponseError:
-        raise NonRecoverableError(EC2ResponseError.body)
+        ctx.logger.error("""(Node: {0}): Error.
+                         Failed to create instance: API returned: {1}."""
+                         .format(ctx.instance.id, EC2ResponseError.body))
+        raise NonRecoverableError('(Node: {0}): Error. Failed to create instance: API returned: {1}.'
+                                  .format(ctx.instance.id, EC2ResponseError.body))
 
     instance_id = reservation.instances[0].id
```
Comment (EarthmanT) [inline ec2/instance.py]: done

File: ec2/instance.py [hunk 9]
```diff
@@ -67,19 +95,38 @@ def start(**kwargs):
 
     instance_id = ctx.instance.runtime_properties['instance_id']
 
+    ctx.logger.info('(Node: {0}): Starting instance.'.format(ctx.instance.id))
+    ctx.logger.debug('(Node: {0}): Attempting to start instance. (Instance id: {1}.)'
+                     .format(ctx.instance.id, instance_id))
+
     try:
         EC2().start_instances(instance_id)
```
Comment (EarthmanT) [inline ec2/instance.py]: If you start an already started machine nothing happens.

File: README.md [hunk 10]
```diff
@@ -3,7 +3,7 @@ cloudify-aws-plugin
 
 A Cloudify Plugin that provisions resources in Amazon Web Services
 
-* Master Branch [![Build Status](https://travis-ci.org/cloudify-cosmo/cloudify-aws-plugin.svg?branch=master)](https://travis-ci.org/cloudify-cosmo/cloudify-aws-plugin)
+* CFY-1879 Branch [![Build Status](https://travis-ci.org/cloudify-cosmo/cloudify-aws-plugin.svg?branch=CFY-1879)](https://travis-ci.org/cloudify-cosmo/cloudify-aws-plugin)
```
Comment (nir0s) [inline README.md]: master...

File: dev-requirements.txt [hunk 11]
```diff
@@ -1,4 +1,3 @@
--e git+https://github.com/cloudify-cosmo/cloudify-dsl-parser@master#egg=cloudify-dsl-parser==3.2a1
--e git+https://github.com/cloudify-cosmo/cloudify-rest-client@master#egg=cloudify-rest-client==3.2a1
--e git+https://github.com/cloudify-cosmo/cloudify-plugins-common@master#egg=cloudify-plugins-common==3.2a1
-nose
```
Comment (nir0s) [inline dev-requirements.txt]: nose.. :)

**Pull request merged**